packages/dashboard-v2/src/components/forms/AddAPIKeyForm.js

```diff
@@ -0,0 +1,114 @@
+import * as Yup from "yup";
+import { forwardRef, useImperativeHandle, useState } from "react";
+import PropTypes from "prop-types";
+import { Formik, Form } from "formik";
+
+import accountsService from "../../services/accountsService";
+
+import { Alert } from "../Alert";
+import { Button } from "../Button";
+import { CopyButton } from "../CopyButton";
+import { TextField } from "../Form/TextField";
+import { CircledProgressIcon, PlusIcon } from "../Icons";
+
+const newAPIKeySchema = Yup.object().shape({
+  name: Yup.string(),
+});
+
+const State = {
+  Pure: "PURE",
+  Success: "SUCCESS",
+  Failure: "FAILURE",
+};
+
+export const APIKeyType = {
+  Public: "public",
+  General: "general",
+};
+
+export const AddAPIKeyForm = forwardRef(({ onSuccess, type }, ref) => {
+  const [state, setState] = useState(State.Pure);
+  const [generatedKey, setGeneratedKey] = useState(null);
+
+  useImperativeHandle(ref, () => ({
+    reset: () => setState(State.Pure),
+  }));
+
+  return (
+    <div ref={ref} className="flex flex-col gap-4">
+      {state === State.Success && (
+        <Alert $variant="success" className="text-center">
+          <strong>Success!</strong>
+          <p>Please copy your new API key below. We'll never show it again!</p>
+          <div className="flex items-center gap-2 mt-4 justify-center">
+            <code className="p-2 rounded border border-palette-200 text-xs selection:bg-primary/30 truncate">
+              {generatedKey}
+            </code>
+            <CopyButton value={generatedKey} className="whitespace-nowrap" />
+          </div>
+        </Alert>
+      )}
+      {state === State.Failure && (
+        <Alert $variant="error">We were not able to generate a new key. Please try again later.</Alert>
+      )}
+      <Formik
+        initialValues={{
+          name: "",
+        }}
+        validationSchema={newAPIKeySchema}
+        onSubmit={async ({ name }, { resetForm }) => {
+          try {
+            const { key } = await accountsService
+              .post("user/apikeys", {
```

Comment:
maybe it's worthwhile turning this into a hook, kinda weird seeing post requests like this in `onSubmit` handlers directly

Reply:
Good idea! 👉  614b7791